packages/next/src/build/babel/loader/get-config.ts

```diff
@@ -90,14 +89,29 @@ function getCacheCharacteristics(
   const isPageFile = pagesDir != null && filename.startsWith(pagesDir)
   const isNextDist = nextDistPath.test(filename)
   const hasModuleExports = source.indexOf('module.exports') !== -1
+  const fileExt = fileExtensionRegex.exec(filename)?.[1] || 'unknown'
+
+  // Compute `hasReactCompiler` as part of the cache characteristics / key,
+  // rather than inside of `getFreshConfig`:
+  // - `isReactCompilerRequired` depends on the file contents
+  // - `node_modules` and `reactCompilerExclude` depend on the file path, which
+  //   isn't part of the cache characteristics
+  let { reactCompilerPlugins, reactCompilerExclude } = loaderOptions
+  reactCompilerPlugins ??= []
+  const hasReactCompiler =
+    reactCompilerPlugins.length !== 0 &&
+    !/[/\\]node_modules[/\\]/.test(filename) &&
+    !reactCompilerExclude?.(filename) &&
+    (await isReactCompilerRequired(filename))
```

Comment:
Is that check worth it for every file? The original idea was to bail out of babel-loader entirely if you just have `reactCompiler` in the config enabled i.e. I would've expected this call to be executed before we even get the config e.g, pass the previously computed value into loader options. We should've just bailed out of getting the config if the React Compiler isn't enabled.

Reply:
Discussed offline:
1. mostly moved so this isn't adding new perf optimization
2. agreed that it's probably not worth it if you're in 2 but can be handled in a follow-up. Babel will parse anyway and adding visitors that don't do anything may not be slower than having to parse and walk the whole source again with SWC. @bgw knows that bailing out of plugin application is worth it if checks are cheap (e.g. `source.includes('class')`

Reply:
Follow-up for the expensive check: https://github.com/vercel/next.js/pull/84103